### PR TITLE
TestCheckpoint: skip on ErrCriuMissingFeatures

### PR DESCRIPTION
**Pull request by AkihiroSuda**

### CODE REVIEW

**Comment by cyphar** [comment]
Do we want to provide information about which features are missing (make an error struct so you can still errors.Is)?

File: libcontainer/criu_linux.go [hunk 1]
```diff
@@ -67,13 +69,13 @@ func (c *Container) checkCriuFeatures(criuOpts *CriuOpts, rpcOpts *criurpc.CriuO
...
-			missingFeatures = true
+			missingFeatures = append(missingFeatures, "LazyPages")
 			logrus.Debugf("CRIU does not support LazyPages")
 		}
 	}
 
-	if missingFeatures {
-		return errors.New("CRIU is missing features")
+	if len(missingFeatures) != 0 {
+		return fmt.Errorf("%w: missing %v", ErrCriuMissingFeatures, missingFeatures)
```
Comment (cyphar) [inline libcontainer/criu_linux.go:78]: Nit: Would it be nicer to make this an error struct?

**Review by cyphar — APPROVED** [review]

File: libcontainer/criu_linux.go (hunk 1 above)
Comment (AkihiroSuda) [reply libcontainer/criu_linux.go:78]: Probably overengineering

Comment (cyphar) [reply libcontainer/criu_linux.go:78]: Yeah, fair enough.

Comment (kolyshkin) [reply libcontainer/criu_linux.go:78]: nit: the error message will look like this

```
criu is missing features: missing [MemTrack LazyPages]
```

I think the second "missing" is redundant.

IOW I suggest having

```go
return fmt.Errorf("%w: %v", ErrCriuMissingFeatures, missingFeatures)
```

**Review by kolyshkin — CHANGES_REQUESTED** [review]
a single nit, otherwise lgtm

File: libcontainer/criu_linux.go (hunk 1 above)
Comment (AkihiroSuda) [reply libcontainer/criu_linux.go]: done

**Comment by AkihiroSuda** [comment]
Addressed the comment

**Comment by lifubang** [comment]
Could you please have a description or link to an issue? To tell us which problem we want to solve.

**Comment by AkihiroSuda** [comment]
> Could you please have a description or link to an issue? To tell us which problem we want to solve.

- https://github.com/opencontainers/runc/pull/4142

On actuated, the kernel is not configured to support criu, so the test has to be skipped, even if criu binary is present.

**Comment by lifubang** [comment]
Thanks, we will always skip this test case in any os which is not configured to support criu, so it will cause us silently skip it if we made a mistake when we modify the CI.
Do we need to add some condition check? For example skip only for some specific OS version?

**Comment by AkihiroSuda** [comment]
> Thanks, we will always skip this test case in any os which is not configured to support criu, so it will cause us silently skip it if we made a mistake when we modify the CI. Do we need to add some condition check? For example skip only for some specific OS version?

We have been checking `criurpc.CriuReq.Features`. Isn't this enough?

**Comment by lifubang** [comment]
I looked into #4142 , it seems that the OS is `ubuntu 22.04` in `actuated`, so maybe we can't skip it only for some os.
My original thought is that we just only skip this test case for the OS in `actuated`.

**Review by lifubang — APPROVED** [review]

**Pull request merged**